src/libbson/src/bson/bson-context.c

```diff
@@ -264,79 +142,236 @@ _bson_context_get_hostname (char *out)
+   uint64_t key[2] = {0};
+   /* The randomness parameters */
+   struct _init_rand_params rand_params;
+
+   /* Init each part of the randomness source: */
+   memset (&rand_params, 0, sizeof rand_params);
+   bson_gettimeofday (&rand_params.time);
+   rand_params.pid = _bson_getpid ();
+   context->gethostname (rand_params.hostname);
+   rand_params.rand_call_counter =
+      bson_atomic_int64_fetch_add (&gRandCounter, 1, bson_memory_order_seq_cst);
+
+   memcpy (key, &rand_params, sizeof key);
+   key[1] = ~key[0];
+
+   /* Hash the param struct */
+   _siphash (&rand_params, sizeof rand_params, key, digest);
+
+   /** Initialize the rand and sequence counters with our random digest */
+   memcpy (context->randomness, digest, sizeof context->randomness);
+   if (init_seq) {
+      memcpy (&context->seq32, digest + 1, sizeof context->seq32);
+      memcpy (&context->seq64, digest + 1, sizeof context->seq64);
+      /* Chop off some initial bits for nicer counter behavior */
+      context->seq32 &= ~UINT32_C (0xf0000f);
+      context->seq64 &= ~UINT64_C (0xf0000f);
+   }
+
+   /* Remember the PID we saw here. This may change in case of fork() */
+   context->pid = rand_params.pid;
 }
 
 static void
 _bson_context_init (bson_context_t *context, bson_context_flags_t flags)
 {
    context->flags = (int) flags;
-   context->oid_set_seq32 = _bson_context_set_oid_seq32;
-   context->oid_set_seq64 = _bson_context_set_oid_seq64;
    context->gethostname = _bson_context_get_hostname;
 
-   if ((flags & BSON_CONTEXT_THREAD_SAFE)) {
-      context->oid_set_seq32 = _bson_context_set_oid_seq32_threadsafe;
-      context->oid_set_seq64 = _bson_context_set_oid_seq64_threadsafe;
-   }
+   context->oid_set_seq32 = _bson_context_set_oid_seq32_threadsafe;
+   context->oid_set_seq64 = _bson_context_set_oid_seq64_threadsafe;
 
-   context->pid = _bson_getpid ();
-   _bson_context_init_random (context, true);
+   _bson_context_init_random (context, true /* Init counters */);
 }
 
 
+/*
+ *--------------------------------------------------------------------------
+ *
+ * _bson_context_set_oid_rand --
+ *
+ *       Sets the process specific five byte random sequence in an oid.
+ *
+ * Returns:
+ *       None.
+ *
+ * Side effects:
+ *       @oid is modified.
+ *
+ *--------------------------------------------------------------------------
+ */
+void
+_bson_context_set_oid_rand (bson_context_t *context, bson_oid_t *oid)
+{
+   BSON_ASSERT (context);
+   BSON_ASSERT (oid);
+
+   if (context->flags & BSON_CONTEXT_DISABLE_HOST_CACHE) {
+      /* User has requested that we check if our PID has changed. This can occur
+       * after a call to fork() */
+      uint64_t now_pid = _bson_getpid ();
+      if (now_pid != context->pid) {
+         _bson_context_init_random (
+            context, false /* Do not update the sequence counters */);
+      }
+   }
+   /* Copy the stored randomness into the OID */
+   memcpy (&oid->bytes[4], &context->randomness, 5);
```

Comment:
```suggestion
   memcpy (&oid->bytes[4], &context->randomness, BSON_CONTEXT_RANDOMNESS_SIZE);
```
Magic number; replace with named constant?

Reply:
Replaced with named constants